CRM/Admin/Page/AJAX.php

```diff
@@ -330,18 +334,25 @@ public static function getTagTree() {
           'style' => $style,
           'class' => 'crm-tag-item',
         ),
-        'children' => (bool) $dao->child,
+        'children' => $hasChildTags,
         'data' => array(
           'description' => (string) $dao->description,
           'is_selectable' => (bool) $dao->is_selectable,
           'is_reserved' => (bool) $dao->is_reserved,
-          'used_for' => $dao->used_for ? explode(',', $dao->used_for) : array(),
+          'used_for' => $usedFor,
           'color' => $dao->color ? $dao->color : '#ffffff',
-          'usages' => (int) $dao->usages,
+          'usages' => civicrm_api3('EntityTag', 'getcount', array(
+            'entity_table' => array('IN' => $usedFor),
+            'tag_id' => $dao->id,
+          )),
         ),
       );
     }
 
+    if (!empty($_REQUEST['is_unit_test'])) {
+      return $result;
+    }
+
```

Comment:
We could remove this hack by making the `CRM_Admin_Page_AJAX::getTagTree()` function a one-line wrapper and move the logic to a BAO function that would do the work and could be more elegantly unit-tested.